Fix LumiCache in AlcaBeamMonitor and OnlineBeamMonitor

francescobrivio opened the pull request:
#### PR description:
Triggered by [this CMSTalk post](https://cms-talk.web.cern.ch/t/issues-with-alcabeammonitor-in-offline-dqm/23945) and as a follow-up of https://github.com/cms-sw/cmssw/issues/37163 and https://github.com/cms-sw/cmssw/issues/39973, @gennai and I implemented the changes suggested by @makortel in https://github.com/cms-sw/cmssw/issues/37163#issuecomment-1061301312.
The changes include:
 - In AlcaBeamMonitor:
    - redefine (and use) the `LuminosityBlockCache`
    - rename the histogram titles from "Scalers" to "Online"
    - Fixed the `Event::getByToken()` try/catch block as suggested by Matti
    - remove the unused `numberOfProcessedLumis_` variable
    - moved the filling of `processedLumis_` to `globalEndLuminosityBlock()` and removed the obsolete `mutable` declaration
- In OnlineBeamMonitor:
   - redefine (and use) the `LuminosityBlockCache`
    - remove the unused `numberOfProcessedLumis_` variable
    - moved the filling of `processedLumis_` to `globalEndLuminosityBlock()` and removed the obsolete `mutable` declaration

@makortel @mmusich please let us know what you think of these changes.

#### PR validation:
~We tested this with wf `138.5` which runs the full `ALCA` reco sequence in step3, using events from multiple lumisections.
I'm not sure if this is the correct way to do it or if some other trick is needed to properly test this, suggestions are welcome.~
See below the discussion about testing this PR properly

#### Backport:
Not a backport. Backports available at:
 - 13_1_X: #41722
 - 13_0_X: #41723

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41700/35575

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @francescobrivio  for master.

It involves the following packages:

- DQM/BeamMonitor (**dqm, db**)


@nothingface0, @emanueleusai, @tvami, @cmsbuild, @saumyaphor4252, @pmandrik, @syuvivida, @tjavaid, @micsucmed, @francescobrivio, @rvenditti can you please review it and eventually sign? Thanks.
@mmusich this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel reviewed (commented):
Thanks @gennai and @francescobrivio! This is the proper way to use `LuminosityBlockCache` in `edm::one` modules.

francescobrivio commented on DQM/BeamMonitor/plugins/AlcaBeamMonitor.cc:
```diff
@@ -224,12 +224,13 @@ void AlcaBeamMonitor::bookHistograms(DQMStore::IBooker& ibooker, edm::Run const&
-  beamSpotsMap_.clear();
+  auto beamSpotInfo = std::make_shared<alcabeammonitor::BeamSpotInfo>();
+  beamSpotInfo->vertices_.clear();
+  beamSpotInfo->beamSpotMap_.clear();
```
done in [f293642](https://github.com/cms-sw/cmssw/pull/41700/commits/f29364240430bfdfc8d1055be62012844e6d7b8d)

francescobrivio commented on DQM/BeamMonitor/plugins/AlcaBeamMonitor.h:
```diff
@@ -28,30 +28,34 @@ class BeamFitter;
+  struct BeamSpotInfo {
+    mutable std::vector<reco::VertexCollection> vertices_;
+    typedef std::map<std::string, reco::BeamSpot> BeamSpotContainer;
+    mutable BeamSpotContainer beamSpotMap_;
```
done in [f293642](https://github.com/cms-sw/cmssw/pull/41700/commits/f29364240430bfdfc8d1055be62012844e6d7b8d)

francescobrivio commented:
In the last push I:
 - applied the suggested comments
 - removed the unused `numberOfProcessedLumis_` from both `AlcaBeamMonitor` and `OnlineBeamMonitor`
 - removed the `mutable` from `processedLumis_` and moved its filling to `globalEndLuminosityBlock()` from both `AlcaBeamMonitor` and `OnlineBeamMonitor`, as suggested in https://github.com/cms-sw/cmssw/issues/37163#issuecomment-1061305897
 - applied the same LumiCache fixes to `OnlineBeamMonitor`

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41700/35579

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #41700 was updated. @nothingface0, @emanueleusai, @tvami, @cmsbuild, @saumyaphor4252, @pmandrik, @syuvivida, @tjavaid, @micsucmed, @francescobrivio, @rvenditti can you please check and sign again.

francescobrivio commented:
I also tested this with:
```
runTheMatrix.py -l 138.5 -j 8 --nStreams 2 --nEvents 1000 --numberEventsInLuminosityBlock 200
```
which gives, e.g.:
<img width="862" alt="Screenshot 2023-05-16 alle 23 33 25" src="https://github.com/cms-sw/cmssw/assets/7822641/9c69f4c2-0eda-4035-8da0-602f141f569e">

To be honest, given the tests parameters I was expecting to see 5 lumisections (1000 evts / 200 evts per LS), but well...better than nothing!
I'm not sure how to instruct the bot to use the same options for testing...any suggestion?

francescobrivio commented:
test parameters:
 - enable = threading

francescobrivio commented:
@cmsbuild please test

 - while waiting for further feedback on the commits and on the correct way to test them, we can still see if the standard tests work in multithreading

mmusich commented:
>  also tested this with:
> runTheMatrix.py -l 138.5 -j 8 --nStreams 2 --nEvents 1000 --numberEventsInLuminosityBlock 200

as far I can tell, when you run in multi-threaded mode (trading `-nStreams 2` with `-t 4`) the plot you show here is changed with respect to the one obtained in a vanilla `CMSSW_13_2_X_2023-05-16-2300`.
![Screenshot from 2023-05-17 12-51-31](https://github.com/cms-sw/cmssw/assets/5082376/d6968510-fd55-438b-9409-844bca2094a3)




On the other hand none of the two is consistent with the observations in the production enviroment (see [CMSTalk post](https://cms-talk.web.cern.ch/t/issues-with-alcabeammonitor-in-offline-dqm/23945), in which apparently only one LS is plotted, https://tinyurl.com/2gtx6s9q). I gather we're not capturing with this command some feature that is running in the Tier-0 jobs.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b8119c/32682/summary.html
**COMMIT**: 39a997ab80c520a80d9785fd5fda2d28cc82213d
**CMSSW**: CMSSW_13_2_X_2023-05-16-2300/el8_amd64_gcc11
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41700/32682/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 9 lines from the logs
 - Reco comparison results: 13 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3222071
 - DQMHistoTests: Total failures: 15
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3222034
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.7560000000000004 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 1000.0,... ): -0.020 KiB       AlcaBeamMonitor/Debug
 - DQMHistoSizes: changed ( 1000.0,... ): -0.009 KiB       AlcaBeamMonitor/Validation
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-05-16-2300+b8119c/57107/triggerResults): no differences found

tvami commented:
test parameters:
  - workflow = 138.5
  - workflow_options =  --nStreams 2 --nEvents 1000 --numberEventsInLuminosityBlock 200

tvami commented:
<img width="788" alt="Screen Shot 2023-05-17 at 07 51 49" src="https://github.com/cms-sw/cmssw/assets/5064110/682398cb-25a7-43de-a691-cf9a30bdcac1">
the bot liked it, so I think this is it

tvami commented:
@cmsbuild , please test

francescobrivio commented:
actually I believe this specific wf is using some of the modules defined here:

https://github.com/cms-sw/cmssw/blob/96423893105e2482c0882f8f0e5fbc61d745605e/Configuration/AlCa/python/autoAlca.py#L60-L66

for which the concurrent lumis are blocked/forbidden. @makortel can you confirm?

mmusich commented:
now running
```
 runTheMatrix.py -l 138.5 -j 8 -t 4 --nEvents 1000 --numberEventsInLuminosityBlock 200 --command='--nConcurrentLumis 2'
```

with this branch I get this


![Screenshot from 2023-05-17 14-52-29](https://github.com/cms-sw/cmssw/assets/5082376/8ebd58bb-39b4-4234-a301-eb6827b57a03)

and things get more consistent (vanilla has only one LS shown), and go in the right direction!

makortel commented:
The `--numberEventsInLuminosityBlock 200` parameter does have any impact here, because the workflow reads existing files and we don't allow reshuffling events into lumis after the fact.

> actually I believe this specific wf is using some of the modules defined here:
> for which the concurrent lumis are blocked/forbidden. @makortel can you confirm?

Indeed the step3 seems to have `PromptCalibProd` and `PromptCalibProdSiPixelAli`, which cause the step3 being run with 1 concurrent lumi regardless of `--command='--nConcurrentLumis 2'`. The override needs to be done in a way that `ConfigBuilder` does not "see" it, which is basically a customise command (`--command='--customise_commands "process.options.numberOfConcurrentLuminosityBlocks=2"'`)  or function, or editing the generated configuration file by hand. But note that the modules currently explicitly not supporting concurrent lumis will misbehave.

In general (two) concurrent lumis are enabled by default for all jobs that use at least 2 streams (with `ConfigBuilder` setting it back to one if some specific generator or AlCa modules are present).

mmusich commented:
> The override needs to be done in a way that ConfigBuilder does not "see" it, which is basically a customise command (--command='--customise_commands "process.options.numberOfConcurrentLuminosityBlocks=2"')

I don't see that in my test.
When I run `runTheMatrix.py -l 138.5 -j 8 -t 4 --nEvents 1000 --numberEventsInLuminosityBlock 200 --command='--nConcurrentLumis 2'` and then dump the step 2 configuration:

```
edmConfigDump step2_RAW2DIGI_L1Reco_RECO_ALCAPRODUCER_DQM_ENDJOB.py > dump.py
```

I see:
```
$ more dump.py | grep currentLumi
    numberOfConcurrentLuminosityBlocks = cms.untracked.uint32(2),
```

makortel commented:
> ump the step 2 configuration

Can you check the step3? (then of course if these modules are used only in the step2, behavior of step3 doesn't matter, but also then the step2 should have concurrent lumis enabled automatically in conjunction of multiple streams)

mmusich commented:
> Can you check the step3? (then of course if these modules are used only in the step2, behavior of step3 doesn't matter, but also then the step2 should have concurrent lumis enabled automatically in conjunction of multiple streams)

indeed

```
$ edmConfigDump step3_ALCAOUTPUT_ALCA.py  > dump2.py
$ more dump2.py | grep currentLumi
    numberOfConcurrentLuminosityBlocks = cms.untracked.uint32(1),
```

but that has no effect on where the DQM module is run.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41700/35586

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #41700 was updated. @nothingface0, @emanueleusai, @tvami, @cmsbuild, @saumyaphor4252, @pmandrik, @syuvivida, @tjavaid, @micsucmed, @francescobrivio, @rvenditti can you please check and sign again.

francescobrivio commented:
> is this ready to be tested?

Might benefit of some commit-squashing, but apart from that i don't have any other updates.

mmusich commented:
> Might benefit of some commit-squashing, but apart from that i don't have any other updates.

standing by for a squash .

francescobrivio committed:
Fix LumiCache in AlcaBeamMonitor

francescobrivio committed:
Fix LumiCache in OnlineBeamMonitor

francescobrivio commented:
> > Might benefit of some commit-squashing, but apart from that i don't have any other updates.
> 
> standing by for a squash .

Thanks Marco! Done in the last push:
 - one commit for all the changes in `AlcaBeamMonitor`
 - one commit for all the changes in `OnlineBeamMonitor`

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41700/35594

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #41700 was updated. @nothingface0, @emanueleusai, @tvami, @cmsbuild, @saumyaphor4252, @pmandrik, @syuvivida, @tjavaid, @micsucmed, @francescobrivio, @rvenditti can you please check and sign again.

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b8119c/32705/summary.html
**COMMIT**: d351d2483eef505816ef39fb8196b0261f720f9a
**CMSSW**: CMSSW_13_2_X_2023-05-17-2300/el8_amd64_gcc11
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41700/32705/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 1829 lines to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3221927
 - DQMHistoTests: Total failures: 40809
 - DQMHistoTests: Total nulls: 19
 - DQMHistoTests: Total successes: 3181077
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.25499999999999956 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 1000.0,... ): -0.020 KiB       AlcaBeamMonitor/Debug
 - DQMHistoSizes: changed ( 1000.0,... ): -0.009 KiB       AlcaBeamMonitor/Validation
 - DQMHistoSizes: changed ( 138.5 ): 0.891 KiB        Hcal/DigiRunHarvesting
 - DQMHistoSizes: changed ( 138.5 ): 0.098 KiB        JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 138.5 ): 0.023 KiB        RPC/DCSInfo
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-05-17-2300+b8119c/57127/triggerResults): no differences found

tvami commented:
> You potentially added 1829 lines to the logs

was this expected?

mmusich commented:
> was this expected?

Given that in the backport tests there is no such increase (
https://github.com/cms-sw/cmssw/pull/41722#issuecomment-1553130713) I tend to conclude it's coming from the non standard test parameters used here.

perrotta commented:
The extra lines in the log are due to the fact the wf 138.5_ExpressCollisions2021 runs with 100 events in the baseline and 1000 events in these tests: is it because of the multithreading enabled?
(Not an issue for the content of this PR, in any case)

francescobrivio commented:
> is it because of the multithreading enabled?

The multithreading and 100 events were part of the non-standard parameters set for the tests (my fault).

@tvami in case you prefer, we can restart the tests after resetting the parameters to the standard values before you sign the PR. Let me know what you prefer, or feel free to do it :)

tvami commented:
>  we can restart the tests 

I dont think we need that, we can just use the backport tests to convince ourselves

tvami commented:
+db
 * PR according to desc
 * tests pass, diffs are understood

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
